Allow loading of const unique_ptr&

rhaschke opened the pull request:
This is the same as https://github.com/pybind/pybind11/pull/2904.
@EricCousineau-TRI: I just migrated to your fork and everything works fine so far. Great work!

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/pybind11/54)
<!-- Reviewable:end -->

rhaschke committed:
Allow loading of const unique_ptr&

rhaschke committed:
fixup! Allow loading of const unique_ptr&

Fix unused parameters.

EricCousineau-TRI commented:
Sweet, glad it worked!
Just to check, can I ask why you ~opened~ closed this PR?

rhaschke commented:
Oh, I might have closed this accidentally when I cleaned up my branches...
I ran into another show-stopping issue in `pydrake` as in the `smart_holder` branch (https://github.com/pybind/pybind11/pull/2911), which is why I decided to proceed on the `smart_holder` branch, which IMO has a more structured code base (at least it was easier to follow for me). That's why I deleted all remaining branches.

EricCousineau-TRI commented:
Makes total sense - thank you for the update!!!
